### PR TITLE
PR: Update Variable Explorer from the kernel

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes

When several clients are connected to a kernel, only the client that sent the execute requests has his variable explorer updated. To solve this, the variable explorer is updated from the kernel instead.

goes with https://github.com/spyder-ide/spyder-kernels/pull/421

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

File: spyder/plugins/ipythonconsole/widgets/shell.py [hunk 1]
```diff
@@ -158,6 +161,8 @@ def __init__(self, ipyclient, additional_options, interpreter_versions,
             'set_debug_state': self.set_debug_state,
             'do_where': self.do_where,
             'pdb_input': self.pdb_input,
+            'update_state': self.update_state,
+
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/shell.py:165]:
```suggestion
```

File: spyder/plugins/ipythonconsole/widgets/shell.py [hunk 2]
```diff
@@ -470,24 +475,17 @@ def get_cwd(self):
...
-            # Frontend sends first
-            return
-        self.call_kernel(callback=self.on_getting_cwd).get_cwd()
-
-    def on_getting_cwd(self, cwd):
+    def update_state(self, state):
         """
-        If necessary, notify that the working directory was changed to other
-        plugins.
+        New state recieved from kernel.
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/shell.py:480]:
```suggestion
        New state received from kernel.
```

File: spyder/plugins/ipythonconsole/widgets/shell.py [hunk 3]
```diff
@@ -470,24 +475,17 @@ def get_cwd(self):
...
         """
-        if cwd != self._cwd:
+        cwd = state.pop("cwd", None)
+        if cwd and self._cwd and cwd != self._cwd:
+            # Only set it self._cwd is already set
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/shell.py:484]:
```suggestion
            # Only set it if self._cwd is already set
```

I think that's what you meant here, right?

File: spyder/plugins/ipythonconsole/widgets/shell.py [hunk 4]
```diff
@@ -470,24 +475,17 @@ def get_cwd(self):
...
             self._cwd = cwd
             self.sig_working_directory_changed.emit(self._cwd)
+        if state:
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/shell.py:487]:
```suggestion

        if state:
```

File: spyder/plugins/ipythonconsole/widgets/shell.py [hunk 5]
```diff
@@ -126,6 +126,9 @@ class ShellWidget(NamepaceBrowserWidget, HelpWidget, DebuggingWidget,
     # To request restart
     sig_restart_kernel = Signal()
 
+    sig_kernel_state = Signal(dict)
+    """The kernel sent a new state to process"""
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/shell.py:130]:
```suggestion
    sig_kernel_state_arrived = Signal(dict)
    """
    A new kernel state, which needs to be processed.
    
    Parameters
    ----------
    state: dict
        Kernel state. The structure of this dictionary is defined in the
        `SpyderKernel.get_state` method of Spyder-kernels.
    """
```

File: spyder/plugins/variableexplorer/widgets/main_widget.py [hunk 6]
```diff
@@ -421,7 +421,7 @@ def create_new_widget(self, shellwidget):
         self._set_actions_and_menus(nsb)
 
         # To update the Variable Explorer after execution
-        shellwidget.executed.connect(nsb.refresh_namespacebrowser)
+        shellwidget.sig_kernel_state.connect(nsb.from_state)
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/main_widget.py:424]:
```suggestion
        shellwidget.sig_kernel_state_arrived.connect(nsb.update_view)
```

File: spyder/plugins/variableexplorer/widgets/main_widget.py [hunk 7]
```diff
@@ -431,7 +431,7 @@ def close_widget(self, nsb):
         nsb.sig_free_memory_requested.disconnect(self.free_memory)
         nsb.sig_start_spinner_requested.disconnect(self.start_spinner)
         nsb.sig_stop_spinner_requested.disconnect(self.stop_spinner)
-        nsb.shellwidget.executed.disconnect(nsb.refresh_namespacebrowser)
+        nsb.shellwidget.sig_kernel_state.disconnect(nsb.from_state)
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/main_widget.py:434]:
```suggestion
        nsb.shellwidget.sig_kernel_state_arrived.disconnect(nsb.update_view)
```

File: spyder/plugins/variableexplorer/widgets/namespacebrowser.py [hunk 8]
```diff
@@ -163,6 +163,14 @@ def refresh_table(self):
         except TypeError:
             pass
 
+    @Slot(dict)
+    def from_state(self, state):
+        """Get from state."""
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/namespacebrowser.py:168]:
```suggestion
    def update_view(self, kernel_state):
        """
        Update namespace view and other properties from a new kernel state.
        
        Parameters
        ----------
        kernel_state: dict
            A new kernel state. The structure of this dictionary is defined in
            the `SpyderKernel.get_state` method of Spyder-kernels.
        """
```

File: spyder/plugins/variableexplorer/widgets/namespacebrowser.py [hunk 9]
```diff
@@ -163,6 +163,14 @@ def refresh_table(self):
...
+        if "namespace_view" in state:
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/namespacebrowser.py:169]:
```suggestion
        if "namespace_view" in kernel_state:
```

File: spyder/plugins/variableexplorer/widgets/namespacebrowser.py [hunk 10]
```diff
@@ -163,6 +163,14 @@ def refresh_table(self):
...
+            self.process_remote_view(state.pop("namespace_view"))
+        if "var_properties" in state:
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/namespacebrowser.py:171]:
```suggestion
        if "var_properties" in kernel_state:
```

**Review by ccordoba12 — COMMENTED** [review]
Thanks @impact27 for this improvement! My review is mostly meant to improve your code's readability, but the implementation looks good to me.

**Comment by impact27** [comment]
I noticed that errors in the frontend would result in pdb hanging. I removed error muting for pdb prompts as otherwise the console would hang without the user understanding why.

**Comment by ccordoba12** [comment]
@impact27, please resync our subrepo since I merged https://github.com/spyder-ide/spyder-kernels/pull/421. Then this should be ready.

**Review by ccordoba12 — APPROVED** [review]
Thanks @impact27! Great work here.

**Pull request merged**